File: Backend/AstroGame.Shared/Models/Stellar/BaseTypes/StellarObject.cs
using AstroGame.Core.Structs;
using AstroGame.Shared.Enums;
using AstroGame.Shared.Models.Resources;
using Newtonsoft.Json;
using System;
using System.Collections.Generic;

namespace AstroGame.Shared.Models.Stellar.BaseTypes
{
    /// <summary>
    /// Anything that can be rendered
    /// </summary>
    public abstract class StellarObject : StellarThing
    {
        protected StellarObject()
        {
        }

        protected StellarObject(StellarSystem parentSystem)
        {
            ParentSystem = parentSystem;
            ParentSystemId = parentSystem.Id;
        }

        public abstract StellarObjectType StellarObjectType { get; }

        public StellarSystem ParentSystem { get; set; }
        public Guid ParentSystemId { get; set; }

        /// <summary>
        /// The average distance from this object to the center it rotates around
        /// NULL means it is the center object
        /// </summary>
        public double? AverageDistanceToCenter { get; set; }

        /// <summary>
        /// The objects relative rotating speed to earth. 1 means 24 hours for a complete rotation
        /// </summary>
        public double RotationSpeed { get; set; }

        /// <summary>
        /// The average temperature in Celsius
        /// </summary>
        public int AverageTemperature { get; set; }

        public string AssetName { get; set; }

        [JsonProperty(Order = -9)] public virtual string Name { get; set; }

        public Coordinates Coordinates { get; set; }

        /// <summary>
        /// The resources stored on this object
        /// </summary>
        public List<ResourceSnapshot> ResourceSnapshots { get; set; }

        [JsonProperty(Order = 100)] public List<StellarObjectResource> Resources { get; set; } = new();
    }
}

File: Frontend/apps/astrogame_app/lib/models/stellar/base_types/stellar_object.dart
import 'package:astrogame_app/communications/converters/guid_converter.dart';
import 'package:astrogame_app/communications/converters/stellar_object_converter.dart';
import 'package:astrogame_app/models/common/coordinates.dart';
import 'package:astrogame_app/models/common/guid.dart';
import 'package:astrogame_app/models/stellar/base_types/stellar_thing.dart';

import 'package:json_annotation/json_annotation.dart';

@GuidConverter()
abstract class StellarObject extends StellarThing {
  // @JsonKey()
  //SingleObjectSystem parentSystem;

  @JsonKey()
  late Guid parentSystemId;

  @JsonKey()
  double? averageDistanceToCenter;

  @JsonKey()
  late double rotationSpeed;

  @JsonKey()
  late int averageTemperature;

  @JsonKey()
  late String assetName;

  @JsonKey()
  late Coordinates coordinates;

  StellarObject();

  factory StellarObject.fromJson(Map<String, dynamic> json) =>
      StellarObjectConverter().fromJson(json);
}
